tcpdevmem.c

```diff
@@ -0,0 +1,119 @@
+
+#include "flow.h"
+#include "lib.h"
+#include "logging.h"
+#include "tcpdevmem_cuda.h"
+#include "tcpdevmem.h"
+#include "thread.h"
+
+#define TEST_PREFIX "ncdevmem"
+#define RETURN_IF_NON_ZERO(cmd)	\
+	ret = (cmd);		\
+	if (ret) return ret;
+
+int driver_reset(const struct options *opts) {
+	char driver_reset_cmd[512];
+	int ret = 0;
+
+	sprintf(driver_reset_cmd, "ethtool --set-priv-flags %s enable-strict-header-split on", opts->tcpd_link_name);
+	RETURN_IF_NON_ZERO(system(driver_reset_cmd));
+
+	sprintf(driver_reset_cmd, "ethtool --set-priv-flags %s enable-strict-header-split off", opts->tcpd_link_name);
+	RETURN_IF_NON_ZERO(system(driver_reset_cmd));
+
+	sprintf(driver_reset_cmd, "ethtool --set-priv-flags %s enable-header-split off", opts->tcpd_link_name);
+	RETURN_IF_NON_ZERO(system(driver_reset_cmd));
+
+	sprintf(driver_reset_cmd, "ethtool --set-rxfh-indir %s equal 16", opts->tcpd_link_name);
+	RETURN_IF_NON_ZERO(system(driver_reset_cmd));
+
+	sprintf(driver_reset_cmd, "ethtool -K %s ntuple off", opts->tcpd_link_name);
+	RETURN_IF_NON_ZERO(system(driver_reset_cmd));
+
+	sprintf(driver_reset_cmd, "ethtool --set-priv-flags %s enable-strict-header-split off", opts->tcpd_link_name);
+	RETURN_IF_NON_ZERO(system(driver_reset_cmd));
+
+	sprintf(driver_reset_cmd, "ethtool --set-priv-flags %s enable-header-split off", opts->tcpd_link_name);
+	RETURN_IF_NON_ZERO(system(driver_reset_cmd));
+
+	sprintf(driver_reset_cmd, "ethtool -K %s ntuple off", opts->tcpd_link_name);
+	RETURN_IF_NON_ZERO(system(driver_reset_cmd));
+
+	sprintf(driver_reset_cmd, "ethtool --set-priv-flags %s enable-max-rx-buffer-size on", opts->tcpd_link_name);
+	RETURN_IF_NON_ZERO(system(driver_reset_cmd));
+
+	sprintf(driver_reset_cmd, "ethtool -K %s ntuple on", opts->tcpd_link_name);
+	RETURN_IF_NON_ZERO(system(driver_reset_cmd));
+
+	return ret;
+}
+
+int install_flow_steering(const struct options *opts, intptr_t buf,
+			  struct thread *t)
+{
+	int q_start = opts->queue_start;
+	int q_num = opts->queue_num;
+	int ret;
+
+	int num_queues = q_start + (t->index % q_num);
+	printf("Bind to queue %i\n", num_queues);
+	struct dma_buf_pages_bind_rx_queue bind_cmd;
+
+	strcpy(bind_cmd.ifname, opts->tcpd_link_name);
+	bind_cmd.rxq_idx = num_queues;
+
+	ret = ioctl(buf, DMA_BUF_PAGES_BIND_RX, &bind_cmd);
+	if (ret < 0)
+	{
+		printf("%s: [FAIL, bind fail queue=%d]\n", TEST_PREFIX,
+		       num_queues);
+		exit(78);
+	}
+
+	/* using t->index below requires 1 thread listening to 1 port
+	 * (see relevant comments in socket.c)
+	 */
+	int src_port = t->index + opts->source_port;
+	int dst_port = t->index + atoi(opts->port);
+
+	char flow_steer_cmd[512];
+	sprintf(flow_steer_cmd,
+		"ethtool -N %s flow-type tcp4 src-ip %s dst-ip %s src-port %i dst-port %i queue %i",
+		opts->tcpd_link_name, opts->tcpd_src_ip, opts->tcpd_dst_ip,
+		src_port, dst_port, num_queues);
+	ret = system(flow_steer_cmd);
+
+	// only running the below ethtool commands after last thread/flow is setup
```

Comment:
I thought we had an open issue to resolve this to reset after every bind?

Reply:
Are you saying you **want** to reset after every bind? So we reset 4x if we have 4 threads/flows?